biz.aQute.bndlib/src/aQute/bnd/cdi/ReferenceDef.java

```diff
@@ -0,0 +1,99 @@
+package aQute.bnd.cdi;
+
+import org.osgi.service.component.annotations.ReferenceCardinality;
+
+import aQute.bnd.osgi.Analyzer;
+import aQute.bnd.osgi.Annotation;
+import aQute.bnd.osgi.Clazz.FieldDef;
+import aQute.bnd.osgi.Descriptors.TypeRef;
+import aQute.bnd.signatures.ClassTypeSignature;
+import aQute.bnd.signatures.FieldSignature;
+import aQute.bnd.signatures.MethodSignature;
+import aQute.bnd.signatures.TypeArgument;
+
+public class ReferenceDef {
+
+	final String				service;
+	ReferenceCardinality		cardinality	= ReferenceCardinality.MANDATORY;
+
+	public ReferenceDef(TypeRef ref) {
+		service = ref.getFQN();
+		cardinality = ReferenceCardinality.MULTIPLE;
+	}
+
+	public ReferenceDef(Annotation reference, FieldDef member, Analyzer analyzer) {
+		Object value = reference.get("value");
+
+		TypeRef ref;
+
+		if (value != null) {
+			ref = (TypeRef) value;
+		}
+		else {
+			String signature = member.getSignature();
+
+			if (signature == null) {
+				signature = member.getDescriptor()
+					.toString();
+			}
+
+			if (signature != null) {
+				ClassTypeSignature type;
```

Comment:
See all my comments above about using the analyzer methods instead of the static `of` methods, about using a resolver to resolve types found in the signatures, and not assuming types can be blindly casted to ClassTypeSignature. Those comments apply here as well.